apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.Tests/RetryTest.cs

```diff
@@ -67,23 +65,31 @@ private static void AssertAttempts(RetryOptions retryOptions, int statusCode, in
 
         var request = service.Buckets.Get("bucket");
         var client = new StorageClientImpl(service);
-        retryOptions ??= RetryOptions.IdempotentRetryOptions;
 
         // Retries for the the failures.
         service.ExpectRequests(request, (HttpStatusCode) statusCode, expectedTotalRetries);
 
-        // Last call is a success.
-        service.ExpectRequest(request, new Bucket());
+        if (success)
+        {
+            // Last call is a success.
+            service.ExpectRequest(request, new Bucket());
 
-        DateTime startTime = DateTime.UtcNow;
-        client.GetBucket("bucket", new GetBucketOptions { RetryOptions = retryOptions });
+            DateTime startTime = DateTime.UtcNow;
+            client.GetBucket("bucket", new GetBucketOptions { RetryOptions = retryOptions });
 
-        Assert.Equal(expectedBackOffs.Count(), messageHandler.AttemptTimestamps.Count());
+            Assert.Equal(expectedBackOffs.Count(), messageHandler.AttemptTimestamps.Count());
 
-        for (int i = 0; i < expectedTotalRetries; i++)
+            for (int i = 0; i < expectedTotalRetries; i++)
```

Comment:
Then we're missing the case of "the code retries a certain number of times but then fails" - either due to different status codes, or due to the limit in Google.Apis. Basically just because we retry doesn't mean we'll eventually succeed.

Reply:
Updated the tests